bigtable-test/bigtable-emulator-maven-plugin/src/main/java/com/google/cloud/bigtable/test/emulator/EmulatorController.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigtable.test.emulator;
+
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.util.Properties;
+
+
+class EmulatorController {
+
+  private final File emulatorPath;
+  private final File portFile;
+  private Process process;
+  private int port;
+  private boolean isStarted;
+
+  EmulatorController(File emulatorPath, File portFile) {
+    this.emulatorPath = emulatorPath;
+    this.portFile = portFile;
+  }
+
+  void start() throws IOException {
+    if (portFile.exists()) {
+      throw new FileAlreadyExistsException(portFile.toString(), null,
+          "Port file already exists. If the emulator crashed, please delete it");
+    }
+
+    port = getFreePort();
+
+    // `gcloud beta emulators bigtable start` creates subprocesses and java doesn't kill process trees
+    // so we have to invoke the emulator directly
+
+    try {
+      process = Runtime.getRuntime().exec(emulatorPath.toString() + " --host=localhost --port=" + port);
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException("cbtemulator could not be found");
+    }
+
+    writeProps();
+
+    isStarted = true;
+  }
+
+  boolean isStarted() {
+    return isStarted;
+  }
+
+  int getPort() {
+    if (port == 0) {
+      throw new IllegalStateException("Emulator hasn't started yet");
+    }
+    return port;
+  }
+
+  void stop() {
+    process.destroy();
+    //noinspection ResultOfMethodCallIgnored
+    portFile.delete();
+  }
+
+  void waitForTermination() throws InterruptedException {
+    process.waitFor();
+  }
+
+  private void writeProps() throws IOException {
+    Properties props = new Properties();
+    props.setProperty("google.bigtable.endpoint.port", Integer.toString(port));
+    props.setProperty("google.bigtable.instance.admin.endpoint.host", "localhost");
+    props.setProperty("google.bigtable.admin.endpoint.host", "localhost");
+    props.setProperty("google.bigtable.endpoint.host", "localhost");
+
+    // emulator doesn't care about projects & instances
+    props.setProperty("google.bigtable.project.id", "fakeproject");
+    props.setProperty("google.bigtable.instance.id", "fakeinstance");
+    // disable authentication
+    props.setProperty("google.bigtable.auth.service.account.enable", "false");
+    props.setProperty("google.bigtable.auth.null.credential.enable", "true");
+
+    Files.createDirectories(portFile.toPath().getParent());
+
+    try (FileOutputStream fout = new FileOutputStream(portFile, false)) {
+      props.store(fout, "Bigtable emulator parameters");
+    }
+  }
+
+  private int getFreePort() throws IOException {
+    try {
+      ServerSocket s = new ServerSocket(0);
```

Comment:
ServerSocket implements AutoClosable, so `try (ServerSocket s = ...) {}` might be stylistically better?